validator/valnode/redis/consumer.go

```diff
@@ -43,46 +44,86 @@ func NewValidationServer(cfg *ValidationServerConfig, spawner validator.Validati
 		consumers[mr] = c
 	}
 	return &ValidationServer{
-		consumers: consumers,
-		spawner:   spawner,
+		consumers:     consumers,
+		spawner:       spawner,
+		streamTimeout: cfg.StreamTimeout,
 	}, nil
 }
 
 func (s *ValidationServer) Start(ctx_in context.Context) {
 	s.StopWaiter.Start(ctx_in, s)
+	// Channel that all consumers use to indicate their readiness.
+	readyStreams := make(chan struct{}, len(s.consumers))
 	for moduleRoot, c := range s.consumers {
 		c := c
+		moduleRoot := moduleRoot
 		c.Start(ctx_in)
-		s.StopWaiter.CallIteratively(func(ctx context.Context) time.Duration {
-			req, err := c.Consume(ctx)
-			if err != nil {
-				log.Error("Consuming request", "error", err)
-				return 0
+		// Channel for single consumer, once readiness is indicated in this,
+		// consumer will start consuming iteratively.
+		ready := make(chan struct{}, 1)
+		s.StopWaiter.LaunchThread(func(ctx context.Context) {
+			for {
+				if pubsub.StreamExists(ctx, c.StreamName(), c.RedisClient()) {
+					ready <- struct{}{}
+					readyStreams <- struct{}{}
+					return
+				}
+				time.Sleep(time.Millisecond * 100)
 			}
-			if req == nil {
-				// There's nothing in the queue.
-				return time.Second
-			}
-			valRun := s.spawner.Launch(req.Value, moduleRoot)
-			res, err := valRun.Await(ctx)
-			if err != nil {
-				log.Error("Error validating", "request value", req.Value, "error", err)
-				return 0
-			}
-			if err := c.SetResult(ctx, req.ID, res); err != nil {
-				log.Error("Error setting result for request", "id", req.ID, "result", res, "error", err)
-				return 0
+		})
+		s.StopWaiter.LaunchThread(func(ctx context.Context) {
+			select {
+			case <-ctx.Done():
+				log.Error("Context done", "error", ctx.Err().Error())
```

Comment:
No need for error. It doesn't really help debugging the source reason (i.e. understanding why context was killed) and just adds a confusing error log.
Info or less.

Reply:
Done.